server/src/main/java/org/elasticsearch/common/geo/GeometryTreeWriter.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.common.geo;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.geo.geometry.Circle;
+import org.elasticsearch.geo.geometry.Geometry;
+import org.elasticsearch.geo.geometry.GeometryCollection;
+import org.elasticsearch.geo.geometry.GeometryVisitor;
+import org.elasticsearch.geo.geometry.Line;
+import org.elasticsearch.geo.geometry.LinearRing;
+import org.elasticsearch.geo.geometry.MultiLine;
+import org.elasticsearch.geo.geometry.MultiPoint;
+import org.elasticsearch.geo.geometry.MultiPolygon;
+import org.elasticsearch.geo.geometry.Point;
+import org.elasticsearch.geo.geometry.Polygon;
+import org.elasticsearch.geo.geometry.Rectangle;
+import org.elasticsearch.geo.geometry.ShapeType;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This is a tree-writer that serializes the
+ * appropriate tree structure for each type of
+ * {@link Geometry} into a byte array.
+ */
+public class GeometryTreeWriter {
+
+    private final GeometryTreeBuilder builder;
+
+    GeometryTreeWriter(Geometry geometry) {
+        builder = new GeometryTreeBuilder();
+        geometry.visit(builder);
+    }
+
+    public BytesRef toBytesRef() throws IOException {
+        BytesStreamOutput output = new BytesStreamOutput();
+        // only write a geometry extent for the tree if the tree
+        // contains multiple sub-shapes
+        boolean prependExtent = builder.shapeWriters.size() > 1;
+        output.writeBoolean(prependExtent);
+        if (prependExtent) {
+            output.writeInt(builder.minLon);
+            output.writeInt(builder.minLat);
+            output.writeInt(builder.maxLon);
+            output.writeInt(builder.maxLat);
```

Comment:
It feels like we could use a `Writable` bounding box class that knows how to read and write itself as well as check for intersections with other bounding boxes. We could then reuse it and in EdgeTreeWriter, which can also become just an `EdgeTree`.